### PR TITLE
Calculate model weights hash in parallel

### PR DESCRIPTION
**Pull request by olpipi**
### Details:
 - Calculate model weights hash in parallel in case of reading model from buffer

### Tickets:
 - CVS-134771

### CODE REVIEW

File: src/inference/src/compilation_context.cpp [hunk 1]
```diff
@@ -121,8 +122,32 @@ std::string ModelCache::compute_hash(const std::string& modelStr,
 
         auto ptr = static_cast<size_t*>(tensor.data());
         size_t size = tensor.get_size() / sizeof(size_t);
-        for (size_t i = 0; i < size; i++)
-            seed = hash_combine(seed, ptr[i]);
+
+        size_t block_size = 100000000;
```
Comment (olpipi) [inline src/inference/src/compilation_context.cpp:126]: 100000000 is optimum for 16GB weights file. In this case we have blocks_num = 22
This value is discussible.

File: src/inference/src/compilation_context.cpp [hunk 2]
```diff
@@ -121,8 +122,32 @@ std::string ModelCache::compute_hash(const std::string& modelStr,
...
+        size_t blocks_num = size / block_size;
+        std::vector<uint64_t> block_hashes(blocks_num + 1);
```
Comment (ilya-lavrenov) [inline src/inference/src/compilation_context.cpp:128]: is it initialized with zeros by default?

File: src/inference/src/compilation_context.cpp (hunk 1 above)
Comment (ilya-lavrenov) [reply src/inference/src/compilation_context.cpp:126]: what if we have 1GB model? does it make sense to perform parallal_for across all threads?
If yes, block_size should be smaller. 

maybe we need to divide total_size by number of threads, but still keep minimal block_size as certain value? In this case, we split work across multiple threads, but if work is small enough, only a few threads will be working.

Comment (olpipi) [reply src/inference/src/compilation_context.cpp]: Updated block_size to make it 1MB or 1000000 / sizeof(size_t)

File: src/inference/src/compilation_context.cpp (hunk 2 above)
Comment (olpipi) [reply src/inference/src/compilation_context.cpp]: fixed

File: src/inference/src/compilation_context.cpp [hunk 3]
```diff
@@ -121,8 +122,35 @@ std::string ModelCache::compute_hash(const std::string& modelStr,
 
         auto ptr = static_cast<size_t*>(tensor.data());
         size_t size = tensor.get_size() / sizeof(size_t);
-        for (size_t i = 0; i < size; i++)
-            seed = hash_combine(seed, ptr[i]);
+
+        // 1MB block size in size_t
```
Comment (ilya-lavrenov) [inline src/inference/src/compilation_context.cpp:126]: maybe we can increase at least to be 10MB?

**Review by p-durandin — APPROVED** [review]

**Review by ilya-lavrenov — APPROVED** [review]

**Pull request merged**